### PR TITLE
New Aggregate Functions

### PR DESCRIPTION
**Pull request by nickwxo**
This pull request implements two new aggregate functions : 

1) The 'sum' function that was here has been replaced. Please see additional comment(s) below for details.

2) A 'date_range' parameter that allows the user to specify a field to be compared with a date range by also specifying either a start_date, an end_date, or both. If only one is specified, it will return all the results with a date after the start date or before the end date accordingly.

Example query : http://localhost:3000/api/crops?end_date=2012-08-31&start_date=2012-08-01&date_range=Plant_Date&District=Ticky%20Ticky - This will return all the crops planted in August of 2008 in the Ticky Ticky district.

@thinkking Please note these features and update the documentation accordingly on a separate branch.

### CODE REVIEW

**Comment by nickwxo** [comment]
http://localhost:3000/api/livestock?sum=Livestock_Count&Livestock_Name=Goats&date_range=Update_Date&start_date=2008-01-01&end_date=2008-12-31&District=Ticky%20Ticky - This will return the total number of goats known to be on farms in Ticky Ticky District that were updated sometime in 2008.

**Comment by nickwxo** [comment]
@matjames007 The changes discussed at last night's meeting have been made so this pull request can be reviewed.

@thinkking You could probably start working on the documentation for these changes in the meantime.

File: routes/resources/crop/crop.js [hunk 1]
```diff
@@ -121,18 +120,88 @@ exports.getAllCrops = function(req, res, next) {
...
-        for (var i = 0;i<crops.length;i++) crops[i] = fakeblock.applyAcl(crops[i], 'get');
-        res.send(crops);
+    };
+
+    /*
+    If we will be returning a count on a particular field we need to modify the attributes of the parameters passed to
+    'findAll' so that it will count the field specified by the user in the 'count' parameter of the query
+     */
+    if (sum != null) {
+        parameters.attributes = [[sequelize.fn('SUM', sequelize.col(sum)), sum]];
```
Comment (matjames007) [inline routes/resources/crop/crop.js]: The Sequelize libraries has several generic functions that can be used applied (`SUM`, `MAX`, `MIN`... etc) that can be used.  Why restrict this endpoint to only `SUM`?  Maybe we could allow for a parameter called `function` which is passed in as a URL parameter and based on that we can apply the function to a particular column.

File: routes/resources/crop/crop.js [hunk 2]
```diff
@@ -121,18 +120,88 @@ exports.getAllCrops = function(req, res, next) {
...
         where: req.query,
         offset: parseInt(offset),
         limit: parseInt(limit)
-    }).then(function(crops) {
-        for (var i = 0;i<crops.length;i++) crops[i] = fakeblock.applyAcl(crops[i], 'get');
-        res.send(crops);
+    };
+
+    /*
+    If we will be returning a count on a particular field we need to modify the attributes of the parameters passed to
```
Comment (matjames007) [inline routes/resources/crop/crop.js]: @nick22891 change the comment to reflect the function being passed.

File: routes/resources/crop/crop.js [hunk 3]
```diff
@@ -121,18 +120,88 @@ exports.getAllCrops = function(req, res, next) {
...
+        delete req.query.start_date;
+        delete req.query.end_date;
+    }
+
+    /*
+     This is the default set of parameters for the findAll function below
+     */
+    var parameters = {
         where: req.query,
         offset: parseInt(offset),
```
Comment (matjames007) [inline routes/resources/crop/crop.js]: if a string is submitted `parseInt` will raise an error and this isn't handled.

File: routes/resources/crop/crop.js [hunk 4]
```diff
@@ -121,18 +120,88 @@ exports.getAllCrops = function(req, res, next) {
...
+        delete req.query.sum;
+    }
+
+    /*
+     We will use the 'date_range' GET parameter to determine whether the user wants to filter by date on a particular field (such as Plant Date
+     in the case of a query on the /crops endpoint.
+     */
+    if (req.query.date_range !== null) {
+        date_range = req.query.date_range;
+        start_date = req.query.start_date;
```
Comment (matjames007) [inline routes/resources/crop/crop.js]: The absence `req.query.start_date` and `req.query.end_date` will throw an error.

**Review by matjames007 — CHANGES_REQUESTED** [review]
Changes needed to handle errors that will be thrown if a url parameter is missing (see in-line comments for details).  Also a suggestion to not be bound by any one aggregate function is made - but not expecting this to be done immediately.

**Comment by nickwxo** [comment]
@matjames007 Please review this pull request again when you can. I've made all the proposed changes with the exception of implementing the other possible functions the library provides. I will create an issue for that now and I've already given @paradoxxjm a list of functions we could easily implement using the library for him to look over and for possible team discussion tomorrow evening.

File: routes/resources/crop/crop.js [hunk 5]
```diff
@@ -121,18 +120,92 @@ exports.getAllCrops = function(req, res, next) {
...
+        Here, we add the start and end dates to the query via the Sequelize library if they exist.
+         */
+        if (start_date != null) date_query.$gte = new Date(start_date);
+        if (end_date != null) date_query.$lte = new Date(end_date);
+        /*
+        And here we add that date query object to the main 'parameters' object that we will pass to the findAll function
+        ONLY IF either a start or end date or both has been specified
+         */
+        if (start_date != null || end_date != null) parameters.where[date_range] = date_query;
+        else res.send("You need to specify a start or end date (or both) to query based on a date range!");
```
Comment (matjames007) [inline routes/resources/crop/crop.js]: Construct the error and send it to the `next()` function

File: routes/resources/crop/crop.js [hunk 6]
```diff
@@ -121,18 +120,92 @@ exports.getAllCrops = function(req, res, next) {
...
+    var parameters = {
         where: req.query,
         offset: parseInt(offset),
         limit: parseInt(limit)
-    }).then(function(crops) {
-        for (var i = 0;i<crops.length;i++) crops[i] = fakeblock.applyAcl(crops[i], 'get');
-        res.send(crops);
+    };
+
+    if (isNaN(parameters.offset) || isNaN(parameters.limit)) res.send("Invalid limit or offset parameter specified. Please ensure they are numeric!");
```
Comment (matjames007) [inline routes/resources/crop/crop.js]: To raise an error try the following:
```
var err = new Error('Invalid Password Given');
err.status = 400;
next(err);
```

**Review by matjames007 — CHANGES_REQUESTED** [review]
Minor changes needed before merging.

**Comment by matjames007** [comment]
We need to specify the appropriate error code.  In this case most of them are bad formatting or 400 errors.

**Comment by nickwxo** [comment]
I have implemented error code handling as per @matjames007 's recommendations.

I have also implemented the aggregate functions differently and they now work as follows : 

6 aggregate functions are currently supported and they all carry out a type of operation on a specific field of the result set. They are : 

(i) sum - This computes the total of all the values from a specific field in the result set.
(ii) avg - This computes the average of all the values from a specific field in the result set.
(iii) min - This returns the lowest of all the values from a specific field in the result set.
(iv) max - This returns the highest of all the values from a specific field in the result set.
(v) var - This returns the statistical variance of all the values from a specific field in the result set.
(vi) stdev - This returns the standard deviation of all the values from a specific field in the result set.

An example of a query would be as follows : 

http://localhost:3000/api/properties?District=Ticky%20Ticky&aggregate_function=avg&field=Property_Size_Ha - This query would give us the average property size in hectares of farms in Ticky Ticky district.

We specify the aggregate function we would like to carry out using the "aggregate_function" parameter and we specify the field we would like to target using the "field" parameter. All other supported query parameters can be used in conjunction with these to enhance specificity.

**Comment by nickwxo** [comment]
@matjames007 Please review when you can.

@thinkking Please update the documentation accordingly.

**Comment by matjames007** [comment]
@nick22891 there are several functions that are shared across all the `/api` resources.  If you can combine the features and have the end result be one function that constructs the `parameters` object based on the `limit` and `offset` values as well as the date ranges supplied and any aggregation information.  This will make for better reuse and a centralized area to maintain all the changes.

**Review by matjames007 — APPROVED** [review]

**Comment by nickwxo** [comment]
I've moved all the duplicate code from each of the /api endpoints to common-util.js and implemented the resulting function in each of the resource endpoints.

@matjames007 Please review again when you can.

**Comment by nickwxo** [comment]
Oh you just did... Lol.

**Pull request merged**